app/controllers/preauthorize_transactions_controller.rb

```diff
@@ -1,263 +1,382 @@
+# coding: utf-8
 class PreauthorizeTransactionsController < ApplicationController
 
   before_filter do |controller|
    controller.ensure_logged_in t("layouts.notifications.you_must_log_in_to_do_a_transaction")
   end
 
-  before_filter :fetch_listing_from_params
   before_filter :ensure_listing_is_open
   before_filter :ensure_listing_author_is_not_current_user
   before_filter :ensure_authorized_to_reply
   before_filter :ensure_can_receive_payment
 
-  BookingForm = FormUtils.define_form("BookingForm", :start_on, :end_on)
-    .with_validations do
-      validates :start_on, :end_on, presence: true
-      validates_with DateValidator,
-                     attribute: :end_on,
-                     compare_to: :start_on,
-                     restriction: :on_or_after
+  IS_POSITIVE = ->(v) {
+    return if v.nil?
+    unless v.positive?
+      {code: :positive_integer, msg: "Value must be a positive integer"}
     end
+  }
 
-  ContactForm = FormUtils.define_form("ListingConversation", :content, :sender_id, :listing_id, :community_id)
-    .with_validations { validates_presence_of :content, :listing_id }
-
-  PreauthorizeMessageForm = FormUtils.define_form("ListingConversation",
-    :content,
-    :sender_id,
-    :contract_agreed,
-    :delivery_method,
-    :quantity,
-    :listing_id
-   ).with_validations {
-    validates_presence_of :listing_id
-    validates :delivery_method, inclusion: { in: %w(shipping pickup), message: "%{value} is not shipping or pickup." }, allow_nil: true
+  PARSE_DATE = ->(v) {
+    return if v.nil?
+    begin
+      TransactionViewUtils.parse_booking_date(v)
+    rescue ArgumentError => e
+      e
```

Comment:
In error case, we need to return something that is not `Date` or `nil` so that the `date` validator can recognize that it's not valid date. We could also return something else like symbol `:error`, no specific reason for the decision to return the exception instead of the symbol.

I'll add a comment explaining this.